### PR TITLE
Improve types of as_sql() and as_<engine>() methods

### PR DESCRIPTION
**Pull request by intgr**
## Related issues

Noticed this in #1243.

### CODE REVIEW

File: django-stubs/contrib/gis/db/models/functions.pyi [hunk 1]
```diff
@@ -1,5 +1,7 @@
 from typing import Any
 
+from django.db.backends.base.base import BaseDatabaseWrapper
+from django.db.backends.mysql.compiler import SQLCompiler
```
Comment (ngnpope) [inline django-stubs/contrib/gis/db/models/functions.pyi]:
```suggestion
from django.db.models.sql.compiler import SQLCompiler, _AsSqlType
```

**Review by ngnpope — CHANGES_REQUESTED** [review]

File: django-stubs/db/models/functions/text.pyi [hunk 2]
```diff
@@ -3,20 +3,18 @@ from typing import Any
...
+    def as_mysql(self, compiler: SQLCompiler, connection: BaseDatabaseWrapper, **extra_context: Any) -> _AsSqlType: ...
 
 class OracleHashMixin:
     def as_oracle(self, compiler: SQLCompiler, connection: BaseDatabaseWrapper, **extra_context: Any) -> _AsSqlType: ...
 
-# Typo: `extra_conteNt`, remains in 4.0
 class PostgreSQLSHAMixin:
     def as_postgresql(
-        self, compiler: SQLCompiler, connection: BaseDatabaseWrapper, **extra_content: Any
+        self, compiler: SQLCompiler, connection: BaseDatabaseWrapper, **extra_context: Any
```
Comment (intgr) [inline django-stubs/db/models/functions/text.pyi:17]: Related: https://github.com/django/django/pull/16437

File: django-stubs/contrib/gis/db/models/functions.pyi (hunk 1 above)
Comment (intgr) [reply django-stubs/contrib/gis/db/models/functions.pyi]: Thanks

File: django-stubs/db/models/functions/text.pyi (hunk 2 above)
Comment (ngnpope) [reply django-stubs/db/models/functions/text.pyi:17]: Damn. That was my typo 😂 Thanks for fixing!

**Comment by intgr** [comment]
This has been sitting for two weeks with no feedback. @ngnpope do you have time to do a review?

**Review by ngnpope — APPROVED** [review]
LGTM.

One that you've missed though:
```
django-stubs/contrib/gis/db/backends/utils.pyi:    def as_sql(self, connection: Any, lookup: Any, template_params: Any, sql_params: Any) -> _AsSqlType: ...
```

**Comment by intgr** [comment]
Rebased and typehinted `SpatialOperator.as_sql` as well. This one has a very different signature from the others, needed some reverse engineering.

File: django-stubs/contrib/gis/db/models/functions.pyi (hunk 1 above)
Comment (ngnpope) [reply django-stubs/contrib/gis/db/models/functions.pyi]: Looks like this has regressed again...

**Comment by intgr** [comment]
Also fixed a failure detected by mypy-self-check, not sure why I didn't notice before.

**Comment by ngnpope** [comment]
> Also fixed a failure detected by mypy-self-check, not sure why I didn't notice before.

https://github.com/typeddjango/django-stubs/pull/1315#discussion_r1086636591 😉

File: django-stubs/contrib/gis/db/models/functions.pyi (hunk 1 above)
Comment (intgr) [reply django-stubs/contrib/gis/db/models/functions.pyi]: Ah I get it now, I did a rebase+force push without pulling the commit added by GitHub "Commit suggestion". My bad

**Pull request merged**